### ISSUE TITLE
multi-monitor behaviour

### ISSUE THREAD

**Issue by koekeishiya** [koekeishiya/chunkwm#45]
Connecting a new monitor and moving windows to it during runtime, seems to work without problems.

The tricky part happens when an already connected monitor, with windows on it, is disconnected / enters sleep or what not.

Soo what happens in my setup at least, is that the windows on the monitor that got disconnected, is moved to the first space of my primary display.

If we now go to the first space, before the monitor is connected, chunkwm will notice these new windows and tile them, and this is fine. When the monitor is reconnected (active) again, MacOS will move these windows back again where they were, however, **chunkwm** has now registered these windows to be tiled on both the first space of the primary display, and whatever this secondary display is, causing problems.

If we do not visit the first desktop of the primary monitor, there is no issues when the monitor reactivates.

Otherwise, there are a couple of things we can do at this point. 

1.
Treat these windows as floating, because some other monitor had them tiled (bad solution, imo). 

2.
We tile the windows on this desktop and untile them from the other monitor (will be forced to remain on this desktop upon monitor reactivation, causes friction, bad solution imo).

3.
We tile the windows on this desktop, cache enough information to allow us to properly untile these windows upon monitor reactivation. Original window trees should just work as if nothing had happened. Ideal solution, but relies on MacOS display notifications to not **** with us.

Another problem is the fact that all other desktops belonging to the removed monitor becomes a part of the primary monitor, and so all our information regarding sizing will be wrong.

**Comment by chrishoage** [comment]
Option 3 seems like the most ideal solution, however it also seems like the most complex which could be error prone (as you mentioned with display notifications not screwing with the system. My experience with a 4k display leads me to think this might be a problem). If it can be stable my vote would be for 3. 

I really wish macOS "merged" the spaces that were on the existing display with the display that was removed. That way windows were not stolen from their space, then placed back if the display was reconnected. The fact that spaces are destroyed when disconnecting a display is pretty annoying behavior (thanks Apple)

**Comment by TheEmpty** [comment]
I agree option 3 would be the best. It sucks when I have to re-arrange my windows after going off to a meeting. Sometimes I try to leave my laptop just so I don't have to re-arrange things.

**Comment by khayyamsaleem** [comment]
Is there any progress on this issue by any chance? Where would we be able to cache the states of the windows and nodes? Because I was thinking, if there was a way to associate the window title with the node, then we can just store that and have it retile based on that association and reconstruct the tree upon reconnection to a monitor.

**Comment by koekeishiya** [comment]
@khayyamsaleem 

Option 3 as mentioned earlier is the planned way of doing this, but I have not yet started on an implementation. I've been using a single-monitor setup lately and have not really had this as a priority.

I do realize that this is a big issue for many users, but the display-system in place in macOS is not exactly pretty to deal with, and so I have not yet bothered to struggle through it.

As for your suggestion regarding association through window title. That is not a long-term solution, as window titles change quite frequently.

For example, whenever I switch the active tab in my browser, the window title changes. This would cause a disconnection between the window and its associated node. Every window has a unique *id* that we can use, and we do already use this to map a given window to a node.

The problem is that (last time I was doing this) we receive spurious monitor notifications, and notifications for "fake" monitors that are created to deal with 4k monitors and so on.

In the current state, we already have the entire window-tree (id + node mappings) intact, the part that makes this a hard problem, is when macOS starts moving windows around, when a secondary monitor is disconnected. This causes our window-tree to mutate (which is as expected, because there are now more windows on some other desktop). When a monitor is connected, some or all of these windows will be moved back, and we need to know which window-trees that needs to be undone.

**Comment by Loyen** [comment]
I've noticed this issue, but it's also an issue when disconnected from the monitor, with windows sometimes overlapping and/or getting bigger margins at the bottom of the screen. I guess it might be related to windows being moved from external monitor to the macbook screen.

**Comment by clevinson** [comment]
Is there an intermediate fix where we can have chunkwm restart if it detects a change in the number of monitors connected (e.g. when a monitor gets connected or discunnected this would trigger a full restart of chunkwm)?

I'd love to have this available as a configuration in my chunkwmrc ASAP, as it seems a proper fix for this is not a high priority.

**Comment by koekeishiya** [comment]
@clevinson 

That should be fairly trivial to do, yes. We would simply add a handler that makes the tiling plugin re-load itself when a monitor is connected / disconnected.
